File: solutions/_740_delete_and_earn/solution.py
from typing import List


class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        if not nums:
            return 0

        points = [0] * (max(nums) + 1)

        for num in nums:
            points[num] += num

        prev = 0
        curr = 0

        for i in range(len(points)):
            temp = curr
            curr = max(curr, prev + points[i])
            prev = temp

        return curr


# https://leetcode.com/problems/delete-and-earn/

File: solutions/_740_delete_and_earn/solution.cpp
#include <vector>
using namespace std;

class Solution
{
public:
    int deleteAndEarn(vector<int> &nums)
    {
        auto max_num = max_element(nums.begin(), nums.end());
        vector<int> points(*max_num + 1, 0);
        for (int num : nums)
        {
            points[num] += num;
        }

        int curr = 0;
        int prev = 0;
        int temp = 0;
        for (int i = 0; i < points.size(); i++)
        {
            temp = curr;
            curr = max(curr, prev + points[i]);
            prev = temp;
        }

        return curr;
    }
};

/*
https://leetcode.com/problems/delete-and-earn/
*/